### PR TITLE
Don't add asyncStorage module to nativeModules if ReactUWP is running without package identity

### PR DESCRIPTION
**Pull request by htpiv**
`AsyncStorageModule` uses `AsyncStorageManager`, which uses `KeyValueStorage`, which uses `StorageFileIO`, which uses `Windows.Storage.StorageFile`, which can only be used from processes with package identity (either UWP apps, or Desktop Bridge apps packaged in an MSIX).

Without this change, creating an instance of `UwpReactInstance` from a classic desktop app would cause an exception to be raised & terminate the app. With it, classic desktop apps can create an instance of `UwpReactInstance`, although `asyncStorage` won't work.

@dotnetjunky, can you please verify that, with this change, your unpackaged version of TodoFeed works correctly?

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3686)

### CODE REVIEW

File: vnext/ReactUWP/Base/UwpReactInstance.cpp [hunk 1]
```diff
@@ -98,6 +100,23 @@
...
 namespace uwp {
 
+namespace {
+
+bool HasPackageIdentity() noexcept {
+  static const bool hasPackageIdentity = []() noexcept {
+    try {
+      auto package = winrt::Windows::ApplicationModel::Package::Current();
+      return true;
+    } catch (...) {
```
Comment (kmelmon) [inline vnext/ReactUWP/Base/UwpReactInstance.cpp]: >... [](start = 13, length = 3)

catch(...) is generally a bad practice.  You should be able to catch a specific type of exception here.

File: vnext/ReactUWP/Base/UwpReactInstance.cpp [hunk 2]
```diff
@@ -231,10 +253,12 @@ std::vector<facebook::react::NativeModuleDescription> GetModules(
       },
       messageQueue);
 
-  modules.emplace_back(
-      "AsyncLocalStorage",
-      []() { return std::make_unique<facebook::react::AsyncStorageModule>(L"asyncStorage"); },
-      std::make_shared<WorkerMessageQueueThread>());
+  if (HasPackageIdentity()) {
```
Comment (kmelmon) [inline vnext/ReactUWP/Base/UwpReactInstance.cpp]: >if (HasPackageIdentity()) { [](start = 2, length = 27)

Please add a comment saying why this check is needed.

File: vnext/ReactUWP/Base/UwpReactInstance.cpp (hunk 1 above)
Comment (htpiv) [reply vnext/ReactUWP/Base/UwpReactInstance.cpp]: I could catch a `winrt::hresult_error const&`, and check that the error is the one I expect, sure.

Comment (htpiv) [reply vnext/ReactUWP/Base/UwpReactInstance.cpp]: Coates reminded me that I should know better than to cause exceptions to be thrown... so I'm doing the crazy incantations to get down to the ABI. I'm not sure why that's so much harder than `winrt::get_abi(winrt::get_activation_factory<Foo, IFooStatics>())->get_Current(...)` but I'm sure there are good reasons...

File: vnext/ReactUWP/Base/UwpReactInstance.cpp [hunk 3]
```diff
@@ -98,6 +101,24 @@
...
+bool HasPackageIdentity() noexcept {
+  static const bool hasPackageIdentity = []() noexcept {
+    auto packageStatics = winrt::get_activation_factory<winrt::Windows::ApplicationModel::IPackageStatics>(
+        winrt::name_of<winrt::Windows::ApplicationModel::Package>());
+    auto abiPackageStatics =
+        static_cast<ABI::Windows::ApplicationModel::IPackageStatics *>(winrt::get_abi(packageStatics));
+    winrt::com_ptr<ABI::Windows::ApplicationModel::IPackage> dummy;
+    return abiPackageStatics->get_Current(reinterpret_cast<ABI::Windows::ApplicationModel::IPackage **>(
+               winrt::put_abi(dummy))) != APPMODEL_ERROR_NO_PACKAGE;
+  }();
```
Comment (dotnetjunky) [inline vnext/ReactUWP/Base/UwpReactInstance.cpp]: I think this function will eventually be used in many other places for the same reason. Can you put it in a separate helper file.

Comment (htpiv) [reply vnext/ReactUWP/Base/UwpReactInstance.cpp]: YAGNI 🙂. We'll move it when/if we need to.

**Pull request merged**